modules/compliance-tests/src/smithy4s/compliancetests/internals/package.scala

```diff
@@ -82,15 +82,33 @@ package object internals {
       Headers(h.toList.flatMap(parseSingleHeader).map(a => a: Header.ToRaw): _*)
     )
 
-  // regex for comma not between quotes as quotes can be used to escape commas in headers
-  private val commaNotBetweenQuotes = ",(?=([^\"]*\"[^\"]*\")*[^\"]*$)"
-
   private def parseSingleHeader(
       kv: (String, String)
   ): List[(String, String)] = {
     kv match {
-      case (k, v) => v.split(commaNotBetweenQuotes).toList.map((k, _))
+      case (k, v) => parseList(v).map((k, _))
     }
+  }
 
+  // splits string into List utilizing a comma that is not between quotes as quotes can be used to escape commas in headers
+  private def parseList(s: String): List[String] = {
+    s.foldLeft((Chain.empty[String], 0, 0, false)) {
+      case ((acc, begin, end, quote), elem) =>
+        elem match {
+          case '"' => (acc, begin, end + 1, !quote)
+          case ',' if !quote && begin < end =>
+            (acc :+ s.substring(begin, end), end + 1, end + 1, quote)
+          case c if c.isWhitespace && begin == end =>
+            (acc, begin + 1, end + 1, quote)
+          case _ =>
+            if (s.length == end + 1)
+              (acc :+ s.substring(begin, end + 1), end + 1, end + 1, quote)
+            else {
+              (acc, begin, end + 1, quote)
+            }
+        }
+    }._1
+      .toList
```

Comment:
It's a bit unclear to me what some of these cases are doing.. I might just be dense, but I think some comment  explaining more what edge cases you are handling here would be helpful

Reply:
you are definitely not dense.
Ill add clear (hopefully) explanations

Reply:
Looks good, thanks for adding 👍